sphinx_exercise/transforms.py

```diff
@@ -0,0 +1,58 @@
+from sphinx.transforms import SphinxTransform
+from sphinx.util import logging
+
+from .nodes import (
+    solution_node,
+    solution_start_node,
+    solution_end_node,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class MergeGatedSolutions(SphinxTransform):
+    """
+    Transform Gated Directives into unified
+    Directives
+    """
+
+    default_priority = 10
+
+    def find_nodes(self, label, node):
+        parent_node = node.parent
+        parent_start, parent_end = None, None
+        for idx1, child in enumerate(parent_node.children):
+            if isinstance(child, solution_start_node) and label == child.get("label"):
+                parent_start = idx1
+                for idx2, child2 in enumerate(parent_node.children[parent_start:]):
+                    if isinstance(child2, solution_end_node):
+                        parent_end = idx1 + idx2
+                        break
+                break
+        if not parent_start or not parent_end:
+            # TODO: raise sphinx.error with details
+            raise Exception
+        return parent_start, parent_end
+
+    def apply(self):
+        for node in self.document.traverse(solution_start_node):
```

Comment:
Maybe it'd be good to do a quick check first and raise some errors for the following conditions:

- Accidentally nesting solution start:

  ````
  ```{solution-start} solution1
  ```
  ```{solution-start} solution2
  ```
  ```{solution-end}
  ```
  ```{solution-end}
  ```
- End solution without start solution
- Start solution without end solution (you have this one documented in the comments I believe)

Reply:
I agree. There needs to be a lot more `robustness` put into the code to check for structural issues and malformed syntax.

Reply:
Thanks @choldgraf that is a good test case 👍